### PR TITLE
MCO-1984: Add Component Readiness tests for boot image skew enforcement

### PR DESCRIPTION
**Pull request by djoshy**
This PR adds 7 e2e tests for the boot image skew enforcement feature:

1. **Manual mode with RHCOSVersion within skew limits** — Sets manual mode with a compliant boot image and verifies the CO's Upgradeable condition is True.
2. **Manual mode with RHCOSVersion outside skew limits** — Sets manual mode with a non-compliant boot image and verifies the CO's Upgradeable condition is False.
3. **Manual mode with OCPVersion within skew limits** — Sets manual mode with a compliant boot image and verifies the CO's Upgradeable condition is True.
4. **Manual mode with OCPVersion outside skew limits** — Sets manual mode with a non-compliant boot image and verifies the CO's Upgradeable condition is False. 
5. **Automatic mode within skew limits**  — Uses the default Automatic mode, backdates a boot image resulting in a reconciliation loop and then verifies that the CO's Upgradeable condition is True.
6. **Automatic mode with degraded controller** — Uses Automatic mode but injects an older boot image with a non-existent user-data secret, causing the boot image controller to degrade. Verifies the CO's Upgradeable condition is False. Then, restores the MachineSet, and verifies that the CO's Upgradeable condition is set back to True.
7. **None mode (opted out)** — Sets None mode to opt out of skew enforcement and verifies the CO's Upgradeable condition is True.

Additional notes: 

- I have also refactoored a few helpers based on review, they have been broken out into a separate commit.

### CODE REVIEW

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 

/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2

rebased to capture changes from https://github.com/openshift/machine-config-operator/pull/5539, and now that https://github.com/openshift/api/pull/2648 has landed, these tests should begin to pass. Unsure if the CRD will show up in the test suites yet, but doing a single run to see

**Comment by djoshy** [comment]
Looks good, running the other gcp shard and the azure variant

/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2

File: test/extended/boot_image_skew.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+var _ = g.Describe("[sig-mco][Suite:openshift/machine-config-operator/disruptive][Serial][Disruptive][OCPFeatureGate:BootImageSkewEnforcement]", func() {
+	defer g.GinkgoRecover()
+
+	var (
+		ManualHappyFixture = filepath.Join("machineconfigurations", "skewenforcement-manual-happy.yaml")
+		ManualSadFixture   = filepath.Join("machineconfigurations", "skewenforcement-manual-sad.yaml")
+		NoneFixture        = filepath.Join("machineconfigurations", "skewenforcement-none.yaml")
+		EmptyFixture       = filepath.Join("machineconfigurations", "machineconfigurations-empty.yaml")
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: Instead of creating 4 templates with hardcoded values, wouldn't the code be more reusable if we create these methods in the MachineConfiguration resource? `SetManualRHCOSSkew(version string)`  (or maybe `SetManualSkew(mode, version string)` ?) and `SetNoneSkew()`  and `RemoveSkew()`

We already have similar methods for .spec.managedBootImages

It would be easier to reuse in future tests.

File: test/extended/boot_image_skew.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+	g.BeforeEach(func() {
+		// Skip on single-node topologies
+		skipOnSingleNodeTopology(oc)
+	})
+
+	g.AfterEach(func() {
+		exutil.By("Resetting MachineConfiguration")
+		// Reset MachineConfiguration between tests
+		applyMachineConfigurationFixture(oc, EmptyFixture)
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: We are not restoring the initial state. Using machineconfiguration.GetSpec at the beginning of the case and machineconfiguration.SetSpec at the end of the case garantees that the initial machineconfguration is restored.

File: test/extended/boot_image_skew.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+	})
+
+	g.It("Verify Manual mode and Upgradeable (Happy case) [apigroup:machineconfiguration.openshift.io]", func() {
+		// Set manual mode with a boot image version that is within skew limits
+		applyMachineConfigurationFixture(oc, ManualHappyFixture)
+
+		// Check machine-config CO upgradeable status, should be set to true
+		upgradeableStatus := extpriv.NewClusterOperator(oc, "machine-config").IsConditionStatusTrue("Upgradeable")
+		o.Expect(upgradeableStatus).Should(o.BeTrue())
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: In my experience Eventually is more robust than Expect, because it will retry in case of an unexpected temporary problem.

Adding a short descriptive message helps a lot when debugging the test if it fails in the CI. "True is not False" message don't give enough information.

```go
o.Eventually(NewClusterOperator(oc, "machine-config").IsConditionStatusTrue, "1m", "10s").WithArguments("Upgradeable").Should(o.BeTrue(), "When automatic skew is configured machine-config CO should not be upgradeable)
``` 

It's closer to natural language

```go
o.Eventually(NewClusterOperator(oc, "machine-config"), "1m", "10s")..Should(HaveConditionField("Upgradeable", "status" "True") , "When automatic skew is configured machine-config CO should not be upgradeable)
```

File: test/extended/boot_image_skew.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		framework.Logf("MachineSet under test: %s", machineSetUnderTest.Name)
+
+		// Set a non-existent user data secret in the machineset's providerSpec; this will cause a boot image controller degrade
+		machineSet := extpriv.NewNamespacedResource(oc, MAPIMachinesetQualifiedName, "openshift-machine-api", machineSetUnderTest.Name)
+		originalUserDataSecret := machineSet.GetOrFail(`{.spec.template.spec.providerSpec.value.userDataSecret.name}`)
+		defer func() {
+			// Restore the original user data secret
+			err := machineSet.Patch("json", `[{"op": "replace", "path": "/spec/template/spec/providerSpec/value/userDataSecret/name", "value": "`+originalUserDataSecret+`"}]`)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}()
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: We can use the MachineSet resource, it contains all the helper functions that you need in this test.

```
machineset := extprive.NewMachineset(oc, "openshift-machine-api", machineSetUnderTest.Name)
originalUserDataSecret, err := machineset.GetUserDataSecretName()
```

We can OrFail a not OrFail function by 

``` go
originalUserDataSecret := extpriv.OrFail[string](machineset.GetUserDataSecretName())
```

File: test/extended/boot_image_skew.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+		nonExistentSecret := "non-existent-user-data-secret"
+		err = machineSet.Patch("json", `[{"op": "replace", "path": "/spec/template/spec/providerSpec/value/userDataSecret/name", "value": "`+nonExistentSecret+`"}]`)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		framework.Logf("Set non-existent user data secret '%s' in MachineSet %s", nonExistentSecret, machineSetUnderTest.Name)
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: Using the MachineSet resource instead of the NamespacedResource we can do like this

``` go
err = machineset.SetUserDataSecret(nonExistentSecret)
```

Expect can handle the error too, so we can delegate the error handling to expect

```go
o.Expect(machineSet.SetUserDataSecret(nonExistentSecret)).To(o.Succeed())
```

File: test/extended/boot_image_skew.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+			switch infra.Status.PlatformStatus.Type {
+			case osconfigv1.GCPPlatformType:
+				restorePatch = `[{"op": "replace", "path": "/spec/template/spec/providerSpec/value/disks/0/image", "value": "` + originalBootImage + `"}]`
+			case osconfigv1.AWSPlatformType:
+				restorePatch = `[{"op": "replace", "path": "/spec/template/spec/providerSpec/value/ami/id", "value": "` + originalBootImage + `"}]`
+			}
+			err := machineSet.Patch("json", restorePatch)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}()
+
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: if we call

```
machineset.SetCoreOSBootimage(myCoreOSBootImage)
```

It will handle all the different paths that depend on the platforms. We need to pay special attention to azure here, since the "coreosbootimage" meaning in azure is the json struct defining the whole coreos boot image (in azure it is not enough the name of the image).

Using this function your test will not have to care about the platform (except when deciding if we run the test or we skip it)

File: test/extended/boot_image_skew.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+			// Restore the original boot image
+			var restorePatch string
+			switch infra.Status.PlatformStatus.Type {
+			case osconfigv1.GCPPlatformType:
+				restorePatch = `[{"op": "replace", "path": "/spec/template/spec/providerSpec/value/disks/0/image", "value": "` + originalBootImage + `"}]`
+			case osconfigv1.AWSPlatformType:
+				restorePatch = `[{"op": "replace", "path": "/spec/template/spec/providerSpec/value/ami/id", "value": "` + originalBootImage + `"}]`
+			}
+			err := machineSet.Patch("json", restorePatch)
+			o.Expect(err).NotTo(o.HaveOccurred())
```
Comment (sergiordlr) [inline test/extended/boot_image_skew.go]: We can restore full spec configurations by writting at the beginning of the test

``` go
defer machineset.SetSpect(machineset.GetSpecOrFail())
```

Then we can modify the resource as we want to without having to worry about what we modified or what we did't modify.

**Comment by sergiordlr** [comment]
We need to take into account that the skew functionality will break the existing test cases configuring machineset bootimages, since those test cases test the Partial and the None configurations and those configurations are forbidden when skew is automatic.

When adding the test cases for the skew functionality, if possible, we should add the code to configure Skew mode None in the already existing bootimages test cases so that they don't collide.

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-2of2

**Comment by djoshy** [comment]
@sergiordlr Thanks for the review! I've addressed all your comments:

- refactor skew enforcement tests to use reusable MachineConfiguration methods (SetManualSkew, SetNoneSkew, RemoveSkew) instead of YAML templates 
- use MachineSet helper methods for manipulating the provider
- use Eventually instead of Expect with descriptive messages for Upgradeable CO checks
- add spec save/restore for MachineConfiguration so original state is not lost

> We need to take into account that the skew functionality will break the existing test cases configuring machineset bootimages, since those test cases test the Partial and the None configurations and those configurations are forbidden when skew is automatic.
>
>When adding the test cases for the skew functionality, if possible, we should add the code to configure Skew mode None in the already existing bootimages test cases so that they don't collide.

Thanks for calling this out! As boot image tests for GCP & AWS are not part of the disruptive suite(they're in origin instead), I overlooked this. I have landed a fix for it via https://github.com/openshift/origin/pull/30712. For Azure/vsphere, they would get affected once boot image updates are enabled by default, so I have added those test changes as part of https://github.com/openshift/machine-config-operator/pull/5583.

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-2of2

File: test/extended-priv/machineconfiguration.go [hunk 8]
```diff
@@ -84,19 +87,19 @@ func (mc MachineConfiguration) GetAllManagedBootImagesResources() ([]string, err
...
 // SetManualSkew configures bootImageSkewEnforcement to Manual mode with the specified mode type and version.
-// mode should be "RHCOSVersion" or "OCPVersion", version is the corresponding version string.
-func (mc MachineConfiguration) SetManualSkew(mode, version string) error {
+// mode should be opv1.ClusterBootImageSpecModeRHCOSVersion or opv1.ClusterBootImageSpecModeOCPVersion.
+func (mc MachineConfiguration) SetManualSkew(mode opv1.ClusterBootImageManualMode, version string) error {
 	logger.Infof("Setting .spec.bootImageSkewEnforcement to Manual mode (%s: %s) on %s", mode, version, mc)
 	var versionField string
 	switch mode {
-	case "RHCOSVersion":
+	case opv1.ClusterBootImageSpecModeRHCOSVersion:
```
Comment (sergiordlr) [inline test/extended-priv/machineconfiguration.go]: We try to avoid importing any api package. 

If we want to use them as constants, I would prefer if we define the constants in the const.go file instead of importing the ones in api/operator/v1, that's the way it is done in all the already existing test cases.

File: test/extended-priv/mco_bootimages.go [hunk 9]
```diff
@@ -1012,3 +1034,50 @@ func setArchitectureAndCheckStatus(clonedMS *MachineSet, machineConfiguration *M
 		"Expected %s not to be BootImageUpdateDegraded.\n%s", machineConfiguration.PrettyString())
 	logger.Infof("No failures are being reported\n")
 }
+
+// SkipUnlessTargetPlatform skips the test if it is not running on any of the target platforms
+func SkipUnlessTargetPlatform(oc *exutil.CLI, platformTypes ...osconfigv1.PlatformType) {
+	infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
```
Comment (sergiordlr) [inline test/extended-priv/mco_bootimages.go]: We try to get all the information via "oc" binary command, the same way a normal user would do it.

This version of the function will do the same as SkipUnlessTargetPlatform, but using the "oc" binary

`skipTestIfSupportedPlatformNotMatched(oc, GCPPlatform, AWSPlatform, AzurePlatform)`

File: test/extended-priv/mco_bootimages.go [hunk 10]
```diff
@@ -1012,3 +1034,50 @@ func setArchitectureAndCheckStatus(clonedMS *MachineSet, machineConfiguration *M
...
+		return
+	}
+	// Special consideration for Azure: we should skip for AzureStack variant
+	if infra.Status.PlatformStatus.Azure != nil && infra.Status.PlatformStatus.Azure.CloudName == osconfigv1.AzureStackCloud {
+		e2eskipper.Skipf("This test does not apply to AzureStack variant within the Azure platform")
+	}
+}
+
+// SkipOnSingleNodeTopology skips the test if the cluster is using single-node topology
+func SkipOnSingleNodeTopology(oc *exutil.CLI) {
```
Comment (sergiordlr) [inline test/extended-priv/mco_bootimages.go]: Same as before, we try to extract all the information via "oc" binary as we would do in our console.

This version of the function will execute the same functionality but using only "oc"

`SkipIfSNO(oc.AsAdmin())`

Nevertheless, we plan to refactor this function, since I think that the new function you used decides how to skip it is better. Our version of IsSNO, is not efficient.

Actually all the "skip..." functions should be refactored/reviewed, at last to place them all in the same file/package. We will do it when the migration finishes (in priv repo we are not allowed to define packages)

File: test/extended-priv/mco_bootimages.go [hunk 11]
```diff
@@ -1012,3 +1034,50 @@ func setArchitectureAndCheckStatus(clonedMS *MachineSet, machineConfiguration *M
...
+	infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+	if infra.Status.ControlPlaneTopology == osconfigv1.SingleReplicaTopologyMode {
+		e2eskipper.Skipf("This test does not apply to single-node topologies")
+	}
+}
+
+// GetRandomMachineSet picks a random machineset present on the cluster
+func GetRandomMachineSet(machineClient *machineclient.Clientset) machinev1beta1.MachineSet {
+	machineSets, err := machineClient.MachineV1beta1().MachineSets("openshift-machine-api").List(context.TODO(), metav1.ListOptions{})
```
Comment (sergiordlr) [inline test/extended-priv/mco_bootimages.go]: Same here, we try to avoid invoking the api directly.

We can leave this function in the "extended" directory.

File: test/extended-priv/boot_image_skew.go [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+	g.AfterEach(func() {
+		exutil.By("Restoring MachineConfiguration to original state")
+		o.Expect(machineConfiguration.SetSpec(originalSpec)).To(o.Succeed())
+	})
+
+	g.It("Verify Manual mode with RHCOSVersion and Upgradeable (Happy case) [apigroup:machineconfiguration.openshift.io]", func() {
+		// Get the current RHCOS version from the coreos-bootimages configmap (guaranteed to be within skew)
+		rhcosVersionWithinSkew := GetRHCOSVersionFromConfigMap(oc)
+		framework.Logf("Using RHCOS version from configmap: %s", rhcosVersionWithinSkew)
```
Comment (sergiordlr) [inline test/extended-priv/boot_image_skew.go]: In all the test cases we use logger.Infof, instead of using framework.Logf

I'm not sure it was a good decision, though. Maybe we should refactor all the logs an use framework.Logf. 

The custom logger allows to create debug messages that only will be printed if the GINKGO_TEST_ENABLE_DEBUG_LOG env var is enabled. As I said, I don't know if it was a good idea, we had to adapt it because the extended fw had some problems with it. However, it would be a good idea if we always use the same logger, and we can change it to framwork.Logf later.

File: test/extended-priv/boot_image_skew.go [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+		// Set manual mode with a boot image version that is within skew limits
+		o.Expect(machineConfiguration.SetManualSkew(opv1.ClusterBootImageSpecModeRHCOSVersion, rhcosVersionWithinSkew)).To(o.Succeed())
+
+		// Wait for the controller to reflect Manual mode in skew enforcement status
+		machineConfiguration.WaitForBootImageSkewEnforcementStatusMode(opv1.BootImageSkewEnforcementModeStatusManual)
```
Comment (sergiordlr) [inline test/extended-priv/boot_image_skew.go]: We try to avoid importing the apis, we can define the constant in the const.go file

File: test/extended-priv/mco_bootimages_skew.go [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+
+		machineSet := NewMachineSet(oc, MachineAPINamespace, machineSetUnderTest.name)
+		// Save and restore full spec to ensure cleanup regardless of what we modify
+		originalMachineSetSpec := machineSet.GetSpecOrFail()
+		defer func() {
+			o.Expect(machineSet.SetSpec(originalMachineSetSpec)).To(o.Succeed())
+		}()
+
+		// Patch the boot image to an older version to trigger an update loop
+		backdatedBootImage := getBackdatedBootImage(oc)
```
Comment (sergiordlr) [inline test/extended-priv/mco_bootimages_skew.go:112]: This existing function does the same ' GetValidUpdateBootImageValue' , but it supports azure and vsphere too.

The name is not a good name, yours is way better. We shoudl rename it in the future.

These 4 functions work together, so that the tests don't have to worry about the platform:
GetValidUpdateBootImageValue
(cpm/ms).SetCoreOsBootImage
(cpms/ms).GetCoreOsBootImage
CheckCurrentOSImageIsUpdated

File: test/extended-priv/boot_image_skew.go [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+		o.Expect(machineConfiguration.RemoveSkew()).To(o.Succeed())
+
+		// Wait for the controller to reflect Automatic mode in skew enforcement status
+		machineConfiguration.WaitForBootImageSkewEnforcementStatusMode(opv1.BootImageSkewEnforcementModeStatusAutomatic)
+
+		// Pick a random machineset to test
+		machineSetUnderTest := NewMachineSetList(oc.AsAdmin(), MachineAPINamespace).GetAllOrFail()[0]
+		framework.Logf("MachineSet under test: %s", machineSetUnderTest.name)
+
+		machineSet := NewMachineSet(oc, MachineAPINamespace, machineSetUnderTest.name)
```
Comment (sergiordlr) [inline test/extended-priv/boot_image_skew.go]: NewMachineSetList(oc.AsAdmin(), MachineAPINamespace).GetAllOrFail()[0] is already returning a machineset, no need to create a new object.

File: test/extended-priv/boot_image_skew.go [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
```
Comment (sergiordlr) [inline test/extended-priv/boot_image_skew.go]: NewMachineSetList(oc.AsAdmin(), MachineAPINamespace).GetAllOrFail()[0] already returns a machineset object

File: test/extended-priv/boot_image_skew.go [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+		framework.Logf("Set backdated boot image '%s' in MachineSet %s to trigger update loop", backdatedBootImage, machineSetUnderTest.name)
+
+		// Verify that the boot image controller has finished processing
+		machineConfiguration.WaitForBootImageControllerComplete()
+
+		// Verify that the boot image controller is degraded
+		machineConfiguration.WaitForBootImageControllerDegradedState(true)
+
+		// Check machine-config CO upgradeable status, should be set to false due to the degrade
+		o.Eventually(NewClusterOperator(oc, "machine-config").IsConditionStatusTrue, "2m", "10s").
```
Comment (sergiordlr) [inline test/extended-priv/boot_image_skew.go]: We are creating the NewClusterOperator machine-config several times, we can define it in the "BeforeEach" sectoin as an item that all tests in the suite can use, or in a var section at the start of the tests so that we only create it once.

File: test/extended-priv/mco_bootimages_skew.go [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+		// Restore machineset to original spec
+		o.Expect(machineSet.SetSpec(originalMachineSetSpec)).To(o.Succeed())
+
+		// Verify that the boot image controller has finished processing
+		machineConfiguration.WaitForBootImageControllerComplete()
+
+		// Verify that the boot image controller is not degraded
+		machineConfiguration.WaitForBootImageControllerDegradedState(false)
+
```
Comment (sergiordlr) [inline test/extended-priv/mco_bootimages_skew.go:173]: Should we check here that the machineconfiguration.status is reporting the right coreos version?

Same question regarding the other  test checking the automatic mode.

File: test/extended-priv/machineconfiguration.go [hunk 19]
```diff
@@ -118,3 +121,39 @@ func (mc MachineConfiguration) RemoveSkew() error {
...
+func (mc MachineConfiguration) WaitForBootImageSkewEnforcementStatusMode(expectedMode opv1.BootImageSkewEnforcementModeStatus) {
+	var mode string
+	o.Eventually(func() (bool, error) {
+		var err error
+		mode, err = mc.GetBootImageSkewEnforcementStatusMode()
+		if err != nil {
+			return false, err
+		}
+		return mode == string(expectedMode), nil
+	}, "2m", "10s").Should(o.BeTrue(), "MachineConfiguration bootImageSkewEnforcementStatus.mode is %s, did not reach expected state: %s", mode, expectedMode)
```
Comment (sergiordlr) [inline test/extended-priv/machineconfiguration.go]: If we define our own constant, we can write

```
o.Eventuall(mc.GetBootImageSkewEnforcementStatusMode, "2m", "10s").Shoul(o.Equal(ConstantSkewManual))
```

File: test/extended-priv/mco_bootimages.go (hunk 10 above)
Comment (djoshy) [reply test/extended-priv/mco_bootimages.go]: ~ack, I will leave a TODO for this one~ actually, I took a shot at refactoring this, let me know what you think!

File: test/extended-priv/mco_bootimages_skew.go (hunk 14 above)
Comment (djoshy) [reply test/extended-priv/mco_bootimages_skew.go:112]: I've renamed it to `getBackdatedBootImage()` and removed the duplicate implementation I had added.

File: test/extended-priv/boot_image_skew.go (hunk 15 above)
Comment (djoshy) [reply test/extended-priv/boot_image_skew.go]: Good catch, thanks! Think I got this mixed up from the original implementation

File: test/extended-priv/mco_bootimages_skew.go (hunk 18 above)
Comment (djoshy) [reply test/extended-priv/mco_bootimages_skew.go:173]: We could, but realistically this is not going to change in error mode or success, since it would be the current version from the start. We would be checking API validation in a way, which seems not necessary.

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-2of2

**Comment by djoshy** [comment]
Updated and rebased per @sergiordlr's review above.

File: test/extended-priv/clusteroperator.go [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+	return &ClusterOperator{Resource: *NewResource(oc, "co", name)}
+}
+
+// NewClusterOperatorList create a ClusterOperatorList struct
+func NewClusterOperatorList(oc *exutil.CLI) *ClusterOperatorList {
+	return &ClusterOperatorList{*NewResourceList(oc, "co")}
+}
+
+// BeUpgradeable returns the gomega matcher to check if a resource is upgradeable or not.
+func BeUpgradeable() types.GomegaMatcher {
```
Comment (sergiordlr) [inline test/extended-priv/clusteroperator.go]: We can place this function in the test/extended-priv/gomega_matchers.go file

Comment (djoshy) [reply test/extended-priv/clusteroperator.go]: Done!

**Comment by sergiordlr** [comment]
These tests will be used for regression testing. They are the implementation of [MCO-1803](https://issues.redhat.com//browse/MCO-1803)

**Comment by djoshy** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-azure-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-gcp-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-vsphere-mco-disruptive-techpreview-2of2 periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-metal-ipi-ovn-dualstack-mco-disruptive-techpreview periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-metal-ipi-ovn-ipv6-mco-disruptive-techpreview periodic-ci-openshift-machine-config-operator-release-4.22-periodics-e2e-metal-ipi-ovn-ipv4-mco-disruptive-techpreview

**Comment by sergiordlr** [comment]
/lgtm

Thank you very much for you help automating the test cases!!
